### PR TITLE
Move docker production build to SBT

### PR DESCRIPTION
**Pull request by swsnr**
Removes `docker/prod` and builds the docker production image with sbt-native-packager instead.

The docker production build now comes down to `docker login; sbt docker:publish` and does not require `prepare-docker-builds` anymore.  SBT completely hides temporary artifacts of the docker build (no more unzipping and copying around) and automatically adds the binary distribution to the docker image.

To inspect the temporary artifacts, run `sbt docker:stage` and inspect `server/target/docker/stage`.  To build the image in the local docker daemon, without publishing to hub, run `sbt docker:publishLocal`.

## Notable differences to the old image

The new image looks like follows

```
FROM openjdk:8-jre
MAINTAINER Frank Rosner
ADD templates /templates
RUN ["mkdir", "/instances"]
RUN ["chown", "-R", "daemon:daemon", "/templates", "/instances"]
RUN ["chmod", "0755", "/instances"]
WORKDIR /opt/docker
ADD opt /opt
RUN ["chown", "-R", "daemon:daemon", "."]
EXPOSE 9000
USER daemon
ENTRYPOINT ["bin/cluster-broccoli"]
CMD ["-Dbroccoli.templates.storage.fs.url=/temlates", "-Dbroccoli.instances.storage.fs.url=/instances"]
```

* It has a `MAINTAINER` tag which supposedly is required for docker hub.
* It runs Broccoli as daemon user, with limited privileges.
* Broccoli is at `/opt/docker` instead `/cluster-broccoli` (that'd be configurable).  Instances are `/opt/docker/instances`, templates at `/opt/docker/templates` by default.
* The image has a `WORKDIR`, ie, you can start the application without specifying the full path as in `bin/cluster-broccoli` as opposed to `/cluster-broccoli/bin/cluster-broccoli`.

## Notable differences in the Travis CI build

* Image tags for branches and travis build numbers are gone.  I could add them back, but do we actually use these tags anywhere?  We still tag with `latest` and the short commit ID.

### CODE REVIEW

**Comment by FRosner** [comment]
Cool! Can we add all necessary information (especially how to inspect the temporary artifacts) to the README file?

File: build.sbt
Comment (FRosner) [inline build.sbt]: should be `/templates`, or?

**Comment by swsnr** [comment]
@FRosner The current README only has information about running the image, not about building/debugging it, so where would you like to add me that information?

**Review by FRosner — COMMENTED** [review]
There is a "building from source" section. Maybe we add some info there? What do you think?

File: build.sbt
Comment (swsnr) [reply build.sbt]: Oh, yes, thanks for catching that.

File: README.md
Comment (FRosner) [inline README.md]: Should be `/templates`

**Review by FRosner — COMMENTED** [review]
I tried running the production image but it breaks.

```
sudo docker run --net host frosner/cluster-broccoli:ff097ca4 -Dbroccoli.templates.storage.fs.url=/templates -Dbroccoli.instances.storage.fs.url=/instances -Dbroccoli.nomad.url=localhost:4646 -Dbroccoli.consul.url=http://localhost:8500
12:59:06.620 INFO akka.event.slf4j.Slf4jLogger - Slf4jLogger started
12:59:07.525 INFO de.frosner.broccoli.services.InstanceService - Nomad/Consul polling frequency set to 1 seconds
12:59:07.528 INFO de.frosner.broccoli.services.InstanceService - broccoli.instances.storage.type=fs
12:59:07.529 INFO de.frosner.broccoli.services.InstanceService - broccoli.instances.storage.fs.url=/instances
12:59:07.531 INFO de.frosner.broccoli.services.FileSystemInstanceStorage - Locking /instances (/instances/.lock)
12:59:07.532 INFO play.api.Play - Application started (Prod)
12:59:07.536 INFO de.frosner.broccoli.services.FileSystemInstanceStorage - readInstances(idFilter) took 1 ms
12:59:07.537 INFO de.frosner.broccoli.services.FileSystemInstanceStorage - readInstances() took 3 ms
12:59:07.650 INFO play.core.server.NettyServer - Listening for HTTP on /0:0:0:0:0:0:0:0:9000
12:59:17.538 INFO de.frosner.broccoli.services.SecurityService - broccoli.auth.session.timeout=3600
12:59:17.552 INFO de.frosner.broccoli.services.SecurityService - broccoli.auth.cookie.secure=true
12:59:17.653 INFO de.frosner.broccoli.services.SecurityService - broccoli.auth.mode=none
12:59:17.777 INFO de.frosner.broccoli.controllers.WebSocketController - New connection 2b4b0ddd-3e00-4891-9fa0-220d2a588f8a by anonymous (Administrator) from 127.0.0.1 at GET /ws
12:59:18.001 ERROR application - 

! @754bc0jb3 - Internal server error, for (GET) [/ws] ->
 
play.api.UnexpectedException: Unexpected exception[IllegalStateException: Templates directory /opt/docker/./templates is not a directory]
	at play.api.http.HttpErrorHandlerExceptions$.throwableToUsefulException(HttpErrorHandler.scala:289)
	at play.api.http.DefaultHttpErrorHandler.onServerError(HttpErrorHandler.scala:220)
	at play.api.GlobalSettings$class.onError(GlobalSettings.scala:160)
	at play.api.DefaultGlobal$.onError(GlobalSettings.scala:188)
	at play.api.http.GlobalSettingsHttpErrorHandler.onServerError(HttpErrorHandler.scala:100)
	at play.core.server.netty.PlayRequestHandler$$anonfun$handle$1.applyOrElse(PlayRequestHandler.scala:129)
	at play.core.server.netty.PlayRequestHandler$$anonfun$handle$1.applyOrElse(PlayRequestHandler.scala:127)
	at scala.concurrent.Future$$anonfun$recoverWith$1.apply(Future.scala:346)
	at scala.concurrent.Future$$anonfun$recoverWith$1.apply(Future.scala:345)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:36)
	at play.api.libs.iteratee.Execution$trampoline$.executeScheduled(Execution.scala:109)
	at play.api.libs.iteratee.Execution$trampoline$.execute(Execution.scala:71)
	at scala.concurrent.impl.CallbackRunnable.executeWithValue(Promise.scala:44)
	at scala.concurrent.impl.Promise$DefaultPromise.tryComplete(Promise.scala:252)
	at scala.concurrent.Promise$class.complete(Promise.scala:55)
	at scala.concurrent.impl.Promise$DefaultPromise.complete(Promise.scala:157)
	at scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:23)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:39)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:415)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: java.lang.IllegalStateException: Templates directory /opt/docker/./templates is not a directory
	at de.frosner.broccoli.templates.DirectoryTemplateSource.loadTemplates(DirectoryTemplateSource.scala:28)
	at de.frosner.broccoli.templates.CachedTemplateSource.refresh(CachedTemplateSource.scala:22)
	at de.frosner.broccoli.templates.CachedTemplateSource.loadTemplates(CachedTemplateSource.scala:18)
	at de.frosner.broccoli.templates.SignalRefreshedTemplateSource.loadTemplates(SignalRefreshedTemplateSource.scala:19)
	at de.frosner.broccoli.services.TemplateService.getTemplates(TemplateService.scala:10)
	at de.frosner.broccoli.controllers.TemplateController$.list(TemplateController.scala:35)
	at de.frosner.broccoli.controllers.WebSocketController$$anonfun$requestToSocket$1.apply(WebSocketController.scala:62)
	at de.frosner.broccoli.controllers.WebSocketController$$anonfun$requestToSocket$1.apply(WebSocketController.scala:32)
	at jp.t2v.lab.play2.auth.BroccoliWebsocketSecurity$class.withSecurity(BroccoliWebsocketSecurity.scala:49)
	at de.frosner.broccoli.controllers.WebSocketController.withSecurity(WebSocketController.scala:18)
	at de.frosner.broccoli.controllers.WebSocketController.requestToSocket(WebSocketController.scala:32)
	at de.frosner.broccoli.controllers.WebSocketController$$anonfun$socket$1.apply(WebSocketController.scala:70)
	at de.frosner.broccoli.controllers.WebSocketController$$anonfun$socket$1.apply(WebSocketController.scala:70)
	at scala.Function1$$anonfun$andThen$1.apply(Function1.scala:52)
	at play.api.mvc.WebSocket$$anonfun$acceptOrResult$1.apply(WebSocket.scala:250)
	at play.api.mvc.WebSocket$$anonfun$acceptOrResult$1.apply(WebSocket.scala:249)
	at play.api.mvc.WebSocket$$anon$1.apply(WebSocket.scala:43)
	at play.core.routing.TaggingInvoker$$anonfun$call$1.apply(HandlerInvoker.scala:48)
	at play.core.routing.TaggingInvoker$$anonfun$call$1.apply(HandlerInvoker.scala:48)
	at play.api.mvc.WebSocket$$anon$1.apply(WebSocket.scala:43)
	at play.core.server.netty.PlayRequestHandler$$anonfun$3.apply(PlayRequestHandler.scala:113)
	at play.core.server.netty.PlayRequestHandler$$anonfun$3.apply(PlayRequestHandler.scala:113)
	at scala.concurrent.impl.Future$PromiseCompletingRunnable.liftedTree1$1(Future.scala:24)
	at scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:24)
	... 6 common frames omitted
```

**Comment by FRosner** [comment]
Ok nvm I just saw that you are still working on it. Ignore the last comment. I will wait for another review request.

**Comment by swsnr** [comment]
@FRosner What commit produced this error message?  Did you try the latest (you shouldn't specify instances and templates dir with this one anymore)?

**Comment by FRosner** [comment]
As mentioned above, it was not the latest commit. As soon as you are finished making changes I will try again with the latest commit.

**Comment by swsnr** [comment]
@FRosner Rebased onto `master` after #265 merged; should be ready to go.

Works on my machine™ :wink:

**Comment by FRosner** [comment]
I tried to run `sbt docker:publishLocal` but got the following error:

```
[info] Successfully built 3fd756449ced
[info] Successfully tagged frosner/cluster-broccoli:0.7.0-SNAPSHOT
[info] Built image frosner/cluster-broccoli:0.7.0-SNAPSHOT
java.lang.IllegalStateException: Ivy file not found in cache for cluster-broccoli-web-ui#cluster-broccoli-web-ui_2.11;0.7.0-SNAPSHOT!
	at sbt.ResolutionCache.getResolvedModuleDescriptor(ResolutionCache.scala:52)
	at org.apache.ivy.core.deliver.DeliverEngine.deliver(DeliverEngine.java:111)
	at org.apache.ivy.Ivy.deliver(Ivy.java:601)
	at sbt.IvyActions$$anonfun$deliver$1.apply(IvyActions.scala:108)
	at sbt.IvyActions$$anonfun$deliver$1.apply(IvyActions.scala:103)
	at sbt.IvySbt$Module$$anonfun$withModule$1.apply(Ivy.scala:156)
	at sbt.IvySbt$Module$$anonfun$withModule$1.apply(Ivy.scala:156)
	at sbt.IvySbt$$anonfun$withIvy$1.apply(Ivy.scala:133)
	at sbt.IvySbt.sbt$IvySbt$$action$1(Ivy.scala:57)
	at sbt.IvySbt$$anon$4.call(Ivy.scala:65)
	at xsbt.boot.Locks$GlobalLock.withChannel$1(Locks.scala:93)
	at xsbt.boot.Locks$GlobalLock.xsbt$boot$Locks$GlobalLock$$withChannelRetries$1(Locks.scala:78)
	at xsbt.boot.Locks$GlobalLock$$anonfun$withFileLock$1.apply(Locks.scala:97)
	at xsbt.boot.Using$.withResource(Using.scala:10)
	at xsbt.boot.Using$.apply(Using.scala:9)
	at xsbt.boot.Locks$GlobalLock.ignoringDeadlockAvoided(Locks.scala:58)
	at xsbt.boot.Locks$GlobalLock.withLock(Locks.scala:48)
	at xsbt.boot.Locks$.apply0(Locks.scala:31)
	at xsbt.boot.Locks$.apply(Locks.scala:28)
	at sbt.IvySbt.withDefaultLogger(Ivy.scala:65)
	at sbt.IvySbt.withIvy(Ivy.scala:128)
	at sbt.IvySbt.withIvy(Ivy.scala:125)
	at sbt.IvySbt$Module.withModule(Ivy.scala:156)
	at sbt.IvyActions$.deliver(IvyActions.scala:103)
	at sbt.Classpaths$$anonfun$deliverTask$1.apply(Defaults.scala:1441)
	at sbt.Classpaths$$anonfun$deliverTask$1.apply(Defaults.scala:1441)
	at scala.Function4$$anonfun$tupled$1.apply(Function4.scala:35)
	at scala.Function4$$anonfun$tupled$1.apply(Function4.scala:34)
	at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
	at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:748)
[error] (webui/*:deliverLocal) java.lang.IllegalStateException: Ivy file not found in cache for cluster-broccoli-web-ui#cluster-broccoli-web-ui_2.11;0.7.0-SNAPSHOT!
```

**Comment by swsnr** [comment]
Can you try again?  And if it still fails, clear your ivy cache? 

I don't think that this error has anything to do with the pull request; I didn't get it, fwiw.

**Review by FRosner — APPROVED** [review]
Tried again, didn't fail. Dunno.

**Pull request merged**
